### PR TITLE
pd: report thread stats in store heartbeat

### PR DESCRIPTION
**Pull request by jiyingtk**
## What have you changed? (mandatory)
Collect thread's cpu usage and disk I/O rates in store's heartbeat:
 - Use tikv_util's HashMap.
 - Temporarily use the kvproto of my branch to support reporting new fields in store's heartbeat.

## What are the type of changes? (mandatory)
- New feature (a change which adds functionality)

## How has this PR been tested? (mandatory)

Test this PR in a TiDB cluster.

## Does this PR affect documentation (docs) or release note? (mandatory)
No.

## Does this PR affect `tidb-ansible` update? (mandatory)
No.

### CODE REVIEW

File: components/tikv_util/src/metrics/threads_dummy.rs [hunk 1]
```diff
@@ -7,9 +7,10 @@ other than Linux. PRs are welcome!
 
 */
 
-use std::collections::HashMap;
 use std::io;
 
+use crate::collections::HashMap;
```
Comment (siddontang) [inline components/tikv_util/src/metrics/threads_dummy.rs]: I find some other places also use std HashMap, can you send a PR later to fix them all?

Comment (jiyingtk) [reply components/tikv_util/src/metrics/threads_dummy.rs]: OK

**Comment by jiyingtk** [comment]
PTAL @rleungx @Luffbee

File: src/raftstore/store/worker/pd.rs [hunk 2]
```diff
@@ -191,6 +297,7 @@ pub struct Runner<T: PdClient> {
     // actually it is the sender connected to Runner's Worker which
     // calls Runner's run() on Task received.
     scheduler: Scheduler<Task>,
+    stats_montor: StatsMonitor,
```
Comment (rleungx) [inline src/raftstore/store/worker/pd.rs]:
```suggestion
    stats_monitor: StatsMonitor,
```

File: src/raftstore/store/worker/pd.rs [hunk 3]
```diff
@@ -172,6 +192,92 @@ impl Display for Task {
...
+        }
+    }
+
+    pub fn start(&mut self) -> Result<(), io::Error> {
+        let (tx, rx) = mpsc::channel();
+        let interval = self.interval;
+        let scheduler = self.scheduler.clone();
+        self.sender = Some(tx);
+        let h = Builder::new()
+            .name("pd-stats-collector".to_owned())
```
Comment (rleungx) [inline src/raftstore/store/worker/pd.rs]: Can we use `thd_name` here?

File: src/raftstore/store/worker/pd.rs [hunk 4]
```diff
@@ -77,6 +84,11 @@ pub enum Task {
     DestroyPeer {
         region_id: u64,
     },
+    StoreInfos {
+        cpu_usages: RecordPairs,
```
Comment (hicqu) [inline src/raftstore/store/worker/pd.rs]: Why not just `Vec`? Maybe in future we will remove `rust-protobuf`.

File: src/raftstore/store/worker/pd.rs [hunk 5]
```diff
@@ -200,7 +307,17 @@ impl<T: PdClient> Runner<T> {
         router: RaftRouter,
         db: Arc<DB>,
         scheduler: Scheduler<Task>,
+        store_heartbeat_interval: u64,
     ) -> Runner<T> {
+        let mut monitor_interval = store_heartbeat_interval / 2;
+        if monitor_interval == 0 {
+            monitor_interval = 1;
```
Comment (hicqu) [inline src/raftstore/store/worker/pd.rs]: It means both `store_heartbeat_interval` and `monitor_interval` are `1s`? Why not just `500ms`?

**Comment by jiyingtk** [comment]
PTAL @hicqu

File: Cargo.toml [hunk 6]
```diff
@@ -143,6 +143,7 @@ default-features = false
 "raft-proto:0.6.0-alpha" = { git = "https://github.com/pingcap/raft-rs", branch = "master" }
 "protobuf:2.8.0" = { git = "https://github.com/nrc/rust-protobuf", branch = "v2.8" }
 "protobuf-codegen:2.8.0" = { git = "https://github.com/nrc/rust-protobuf", branch = "v2.8" }
+"https://github.com/pingcap/kvproto.git#kvproto:0.0.2" = { git = "https://github.com/jiyingtk/kvproto", branch = "add-fields-in-store-heartbeat" }
```
Comment (hicqu) [inline Cargo.toml]: Please update it.

File: src/raftstore/store/worker/pd.rs [hunk 7]
```diff
@@ -200,7 +307,14 @@ impl<T: PdClient> Runner<T> {
         router: RaftRouter,
         db: Arc<DB>,
         scheduler: Scheduler<Task>,
+        store_heartbeat_interval: u64,
     ) -> Runner<T> {
+        let interval = Duration::from_secs(store_heartbeat_interval).div_f64(2.0);
```
Comment (hicqu) [inline src/raftstore/store/worker/pd.rs]: There is `impl Div<u32> for Duration` in std lib. BTW make `2.0` as a constant is better.

**Comment by hicqu** [comment]
rest LGTM. PTAL @rleungx thanks!

**Pull request merged**